include/lmms_math.h

```diff
@@ -96,6 +96,57 @@ static void roundAt(T& value, const T& where, const T& stepSize)
 	}
 }
 
+//! Uses `x * y + z` or `std::fma` - whichever is faster
+template<typename T>
+inline T fastFma(T x, T y, T z);
```

Comment:
Is this worth having? It's not like a replace for `std::pow` or `std::sqrt` or something, where we'd have to call a library function anyway. It instead replaces the normal arithmetic expression `x * y + z`, which I find considerably easier to read when part of more complex expressions, and which contributors are less likely to go looking for a function to replace.

Also, I can't actually get a compiler to generate faster code with this function. I'll go over the specific conditions for each compiler, but they all generate one of the following sequences of instructions with optimisations enabled (this is for AMD64 with `T = double`):
```asm
; Seperate multiply and add
mulsd   xmm0, xmm1
addsd   xmm0, xmm2

; Fused multiply-add
vfmadd213sd xmm0, xmm1, xmm2 ; GCC prefers `vfmadd132sd xmm0, xmm2, xmm1`

; Tail call into C library
jmp     fma
```
GCC and Clang behave very similarly. For target architectures below `x86-64-v3`, the plain expression `x * y + z` compiles to separate instructions, and `std::fma(x, y, z)` calls into the C library. Neither compiler defines `FP_FAST_FMA`. For `-march=x86-64-v3` and above, both of these compile to the single fused instruction. GCC defines `FP_FAST_FMA` but Clang does not; however, it doesn't matter as both alternatives produce the same code.

MSVC's equivalent of `-march=x86-64-v3` is `/arch:AVX2`. Below this, the behaviour is exactly the same as GCC and Clang. For `/arch:AVX2` and above, the behaviour is a bit different. While `std::fma(x, y, z)` always compiles to the fused instruction, MSVC will generate separate instructions for `x * y + z` unless `/fp:fast` (equivalent to `-ffast-math`) is passed. However, `FP_FAST_FMA` is never defined. This seems inconsistent to me - if the fused version is faster, then `FP_FAST_FMA` should be defined, and if it isn't, then `/fp:fast` shouldn't result in its generation.

Similar results are obtained for x86, AArch64, and AArch32, and for `T = float`: `std::fma(x, y, z)` is either slower than `x * y + z`, or generates identical code. The only exception is under MSVC, which only generates identical code with `/fp:fast`, but this makes no difference here since it doesn't define `FP_FAST_FMA`.

Reply:
Just in case it is not already known: one can test these things in Godbolt's Compiler Explorer which can be found at https://godbolt.org/.

Reply:
I've removed it now, replacing calls to `fastFma` with `x * y + z`.